src/nvim/fileio.c

```diff
@@ -493,22 +493,14 @@ readfile (
     curbuf->b_flags &= ~(BF_NEW | BF_NEW_W);
   }
 
-  /*
-   * Check readonly by trying to open the file for writing.
-   * If this fails, we know that the file is readonly.
-   */
-  file_readonly = FALSE;
+  // Check readonly.
+  file_readonly = false;
   if (!read_buffer && !read_stdin) {
-    if (!newfile || readonlymode) {
-      file_readonly = TRUE;
-    } else if ((fd = os_open((char *)fname, O_RDWR, 0)) < 0) {
-      // opening in readwrite mode failed => file is readonly
-      file_readonly = TRUE;
-    }
-    if (file_readonly == TRUE) {
-      // try to open readonly
-      fd = os_open((char *)fname, O_RDONLY, 0);
+    if (!newfile || readonlymode || !(perm & 0222)
```

Comment:
@Shougo Thanks for noticing that. I'm working on a fix. Even the Vim version of readfile() seems rather suspect. 

Under the conditions `(fd >= 0 && (read_stdin || read_buffer))`, the following code uses uninitialized `perm`:

https://github.com/vim/vim/blob/30e12d259ee78272359f9da2655d0593a4f6a626/src/fileio.c#L2449